### DISCUSSION TITLE
[Windows] Extending database size and OOM

### DISCUSSION THREAD

**Discussion by oleg-kiriyenko** [erthink/libmdbx#354] [Q&A] [answered]
The problem is with intensive writing into the database. When it's true two processes fail (killed?) (reader and writer). I think that I do not understand correcly how database is extended and OOM killer works.

1. Previous version (0.5) had  MDBX_MAP_RESIZED. But new version doesn't. How extending of database is handled now?
-------------------

2. Description of oom_killer function is not clear enough
----------------------

* Returns -1 on failure (reader is not killed)

Does that mean other options will kill reader? Or we should do it manually in killer function?

* 0 should wait or retry

That will do nothing with reader?

* 1 drop reader txn-lock (reading-txn was aborted),

How drop lock is done? In reallity reader is killed after calling mdbx_txn_reset().

>1 drop reader registration (reader process was killed)

What does that mean? That we have killed it in killer function?

**Comment by erthink** [comment, marked as answer]
> 1. Previous version (0.5) had  MDBX_MAP_RESIZED. But new version doesn't. How extending of database is handled now?

The `MDBX_MAP_RESIZED` was replaced with [`MDBX_UNABLE_EXTEND_MAPSIZE`](https://libmdbx.dqdkfa.ru/group__c__err.html#ggab5ba1469eef28a2e09ed243b608b11cea10c8033f7aeb8809747cc517dce311e0). This is reasonable even breaking source-level compatibility, since semantic of `MDBX_MAP_RESIZED` has been inherited from LMDB and obsolete a long time ago.

> 2. Description of oom_killer function is not clear enough

_libmdbx_ doesn't kill anything itself. The [MDBX_hsr_func](https://libmdbx.dqdkfa.ru/group__c__err.html#gaf6f9648717a250998227f34816a37d3e) callback (aka "oom_killer") is  optional. If such a function was provided, _libmdbx_ will call it as described, and the return value must match what was actually done by your function.

**Comment by oleg-kiriyenko** [comment]
1) And what is the difference between -1 and 0? How libmdbx will react to this codes?

2) If we report 1 does that mean that we somehow forced other process to release lock (maybe using some IPC)? liblmdx will not check anything and think that lock was released?

**Comment by erthink** [comment]
> 1. And what is the difference between -1 and 0? How libmdbx will react to this codes?

Sometimes it's better to look at the source code: https://github.com/erthink/libmdbx/blob/abf38e97cce09f2fb6f62c1d0dcde301ff431137/src/elements/core.c#L16623-L16639

And some usage example:  https://github.com/erthink/libmdbx/blob/abf38e97cce09f2fb6f62c1d0dcde301ff431137/test/test.cc#L81-L100

I.e. if return code is...
   - `less than zero`: _libmdbx_ will exit from oom-handle loop and try to extend DB  or return correcponding error code.
   - `equal to zero`: _libmdbx_ will continue oom-loop, rescan readers lock table and (depending on the situation - there is space or not) will exit the oom-loop without further DB extension, or call your function again.
   - `great than zero`: please see [description](https://github.com/erthink/libmdbx/blob/abf38e97cce09f2fb6f62c1d0dcde301ff431137/mdbx.h#L3409-L3415).

-----

> 2. If we report 1 does that mean that we somehow forced other process to release lock (maybe using some IPC)?

Yes, returning a value > 0 means this exactly.

> liblmdx will not check anything and think that lock was released?

Yes, will not check.

However, there is a difference when returning `== 1` or `>= 2`:
   - `== 1`: Means that the reader thread was not killed and will not continue reading data within a currently (its own) running transaction, but will call (immediatelly or later) `mdbx_txn_abort()` or `mdbx_txn_reset()`. In this case _libmdbx_ will [clear only `mr_txnid` int the slot by compare-and-swap](https://github.com/erthink/libmdbx/blob/abf38e97cce09f2fb6f62c1d0dcde301ff431137/src/elements/core.c#L16631) since reader thread run asynchronously and may executing alredy new (restarted) transaction (with new txnid) at the moment.
  - `>= 2`: Means that the reader thread (or a whole process) was killed and will NOT call `mdbx_txn_abort()` or `mdbx_txn_reset()`. In this case _libmdbx_ [clear whole slot unconditionaly](https://github.com/erthink/libmdbx/blob/abf38e97cce09f2fb6f62c1d0dcde301ff431137/src/elements/core.c#L16633-L16635).

**Comment by erthink** [comment]
@oleg-kiriyenko, please close this issue if there are no more questions.
